### PR TITLE
[Merged by Bors] - feat: replace have? correctly

### PR DESCRIPTION
**Pull request by fpvandoorn**
* `have? using h` now replaces the whole tactic with the suggestion, not just the `have?`
* Allow `have!?` as alternative syntax for `have?!` (similar for `rw!?` and `exact!?` -- the latter has not been implemented yet)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/Propose.lean [hunk 1]
```diff
@@ -123,14 +123,17 @@ elab_rules : tactic
         throwError "propose could not find any lemmas using the given hypotheses"
       -- TODO we should have `proposals` return a lazy list, to avoid unnecessary computation here.
       for p in proposals.toList.take 10 do
-        addHaveSuggestion tk (← inferType p.2) p.2
+        addHaveSuggestion tk (← inferType p.2) p.2 (← getRef)
```
Comment (kmill) [inline Mathlib/Tactic/Propose.lean:126]: Maybe you could put `let stx ← getRef` right before the `getMainGoal` line to make it easier to reason about whether `getRef` is getting the right thing.

Comment (digama0) [reply Mathlib/Tactic/Propose.lean:126]: I believe `getRef` reads from the read-only state of the monad, so the only thing that can cause issues is if you use a combinator (with the obvious one to watch out for being `withRef`).

**Comment by semorrison** [comment]
Kyle's suggestion looks good, otherwise LGTM.

bors d+

**Comment by fpvandoorn** [comment]
I forgot to merge this

bors merge

**Pull request closed without merging**